crunchers/test.cc

```diff
@@ -201,16 +199,449 @@ TEST(CTFOCrunchersTest, ActiveUsersCruncherRemoteTest) {
+  EXPECT_EQ(1.0, top_cards.value[0][1].rate);
+  // Cruncher #2 (2ms time window).
+  // Card #3: rate = (TFU (1x) + FAV (1x)) / SEEN (1x) = 2.
+  // Card #1: rate = SKIP (1x) / SEEN (1x) = 1.
+  // Card #2: rate = CTFO (1x) / SEEN (0x) = 0.
+  ASSERT_EQ(3u, top_cards.value[1].size());
+  EXPECT_EQ(3u, top_cards.value[1][0].cid);
+  EXPECT_EQ(calculators[1](top_cards.value[1][0]), top_cards.value[1][0].rate);
+  EXPECT_EQ(2.0, top_cards.value[1][0].rate);
+  EXPECT_EQ(1u, top_cards.value[1][1].cid);
+  EXPECT_EQ(calculators[1](top_cards.value[1][1]), top_cards.value[1][1].rate);
+  EXPECT_EQ(1.0, top_cards.value[1][1].rate);
+  EXPECT_EQ(2u, top_cards.value[1][2].cid);
+  EXPECT_EQ(calculators[1](top_cards.value[1][2]), top_cards.value[1][2].rate);
+  EXPECT_EQ(0.0, top_cards.value[1][2].rate);
+
+  // Add three more events for Card #3: SEEN, CTFO and UNKNOWN (the last one should not affect the ratings).
+  current::time::SetNow(std::chrono::microseconds(3001));
+  ios_event.event = "SEEN";
+  ios_event.user_ms = std::chrono::duration_cast<std::chrono::milliseconds>(current::time::Now());
+  log_entry.server_us = current::time::Now();
+  log_entry.event = ios_event;
+  local_stream.Publish(CTFO_Local::CTFOLogEntry(log_entry));
+  current::time::SetNow(std::chrono::microseconds(3002));
+  ios_event.event = "CTFO";
+  ios_event.user_ms = std::chrono::duration_cast<std::chrono::milliseconds>(current::time::Now());
+  log_entry.server_us = current::time::Now();
+  log_entry.event = ios_event;
+  local_stream.Publish(CTFO_Local::CTFOLogEntry(log_entry));
+  current::time::SetNow(std::chrono::microseconds(3003));
+  ios_event.event = "UNKNOWN";
+  ios_event.user_ms = std::chrono::duration_cast<std::chrono::milliseconds>(current::time::Now());
+  log_entry.server_us = current::time::Now();
+  log_entry.event = ios_event;
+  local_stream.Publish(CTFO_Local::CTFOLogEntry(log_entry));
+  while (GetTopCards(base_url, 0).timestamp < current::time::Now()) {
+    std::this_thread::yield();
+  }
+
+  top_cards = GetTopCards(base_url);
+  EXPECT_EQ(3003u, top_cards.timestamp.count());
+  ASSERT_EQ(2u, top_cards.value.size());
+  // Cruncher #1 (1ms time window).
+  // Card #3: rate = SEEN (1x) + CTFO (1x) = 2.
+  // Events TFU (1x), FAV (1x) and SEEN (1x) for Card #3 is out of the time window.
+  // Cards #1 and #2 are out of the time window.
+  ASSERT_EQ(1u, top_cards.value[0].size());
+  EXPECT_EQ(3u, top_cards.value[0][0].cid);
+  EXPECT_EQ(calculators[0](top_cards.value[0][0]), top_cards.value[0][0].rate);
+  EXPECT_EQ(2.0, top_cards.value[0][0].rate);
+  // Cruncher #2 (2ms time window).
+  // Card #3: rate = (TFU (1x) + FAV (1x) + CTFO (1x)) / SEEN (2x) = 1.5.
+  // Cards #1 and #2 are out of the time window.
+  ASSERT_EQ(1u, top_cards.value[1].size());
+  EXPECT_EQ(3u, top_cards.value[0][0].cid);
+  EXPECT_EQ(calculators[1](top_cards.value[1][0]), top_cards.value[1][0].rate);
+  EXPECT_EQ(1.5, top_cards.value[1][0].rate);
+}
```

Comment:
Of course. I suggest to do it for all of the crunchers tests at once in a separate pull request.